File: src/audio/whisper_daemon_v2.py
#!/usr/bin/env python3
"""
Whisper Daemon V2 — GPU-accelerated streaming STT for Echo Mode.

Upgrades over V1:
- GPU support via CUDA ctranslate2 (cublas/cudnn auto-loaded from pip)
- faster-whisper with configurable model/device/compute_type
- Streaming partial transcripts via chunk_length parameter
- Health check command
- Structured JSON logging to stderr
- Graceful fallback: CUDA → CPU if GPU fails

Protocol (stdin → stdout, one JSON per line):
  {"op": "transcribe", "path": "/tmp/xxx.wav"}
    → {"ok": true, "text": "...", "partial": false}
    → {"ok": true, "text": "...", "partial": true}  (if streaming)

  {"op": "transcribe_stream", "path": "/tmp/xxx.wav"}
    → {"ok": true, "text": "partial1", "partial": true}
    → {"ok": true, "text": "partial2", "partial": true}
    → {"ok": true, "text": "final text", "partial": false}

  {"op": "ping"}
    → {"ok": true, "pong": true, "model": "base", "device": "cuda"}

  {"op": "health"}
    → {"ok": true, "healthy": true, "gpu": true, "vram_mb": 766}

  {"op": "quit"}
    → process exits

Environment:
  CUDA libs must be on PATH. Set WHISPER_DAEMON_DEVICE=cuda|cpu
  and WHISPER_DAEMON_MODEL=base|small|large-v3-turbo etc.
"""

import sys
import os
import json
import time
import traceback

# ─── GPU Library Path Setup ──────────────────────────────────────────────
# The cublas/cudnn DLLs installed via pip need to be on PATH for ctranslate2
# to find them. We add them here before any ctranslate2 import.
_VENV_NVIDIA = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))),
    "..", "..", "AppData", "Local", "hermes", "hermes-agent", "venv", "Lib",
    "site-packages", "nvidia"
)
# Also check standard site-packages location
try:
    import site
    for sp in site.getsitepackages():
        _nv = os.path.join(sp, "nvidia")
        if os.path.isdir(_nv):
            _VENV_NVIDIA = _nv
            break
except:
    pass

# Add CUDA DLL directories to PATH
for sub in ["cublas", "cudnn", "cuda_runtime"]:
    bindir = os.path.join(_VENV_NVIDIA, sub, "bin")
    if os.path.isdir(bindir):
        try:
            os.environ["PATH"] = bindir + os.pathsep + os.environ.get("PATH", "")
            # Also try Windows DLL directory API
            if hasattr(os, "add_dll_directory"):
                os.add_dll_directory(bindir)
        except:
            pass

# ─── Configuration ────────────────────────────────────────────────────────
MODEL_NAME = os.environ.get("WHISPER_DAEMON_MODEL", "base")
DEVICE = os.environ.get("WHISPER_DAEMON_DEVICE", "cuda")
COMPUTE_TYPE = os.environ.get("WHISPER_DAEMON_COMPUTE", "int8")
LANGUAGE = os.environ.get("WHISPER_DAEMON_LANG", "en")
BEAM_SIZE = 1
STREAM_CHUNK_LENGTH = 1.0  # seconds, for streaming mode

def log(msg: str):
    """Structured logging to stderr."""
    print(f"[WhisperDaemonV2] {msg}", file=sys.stderr, flush=True)

def emit(obj: dict):
    """Emit a JSON object on stdout (one line)."""
    print(json.dumps(obj), flush=True)

def load_model():
    """Load faster-whisper model with GPU support and CPU fallback."""
    global model, device, compute_type, using_gpu

    from faster_whisper import WhisperModel

    # Try GPU first if device != "cpu"
    using_gpu = False
    if DEVICE != "cpu":
        try:
            log(f"Loading model '{MODEL_NAME}' on CUDA ({COMPUTE_TYPE})...")
            t0 = time.time()
            model = WhisperModel(MODEL_NAME, device="cuda", compute_type=COMPUTE_TYPE)
            load_time = time.time() - t0

            # Warmup — run a tiny transcribe to ensure GPU is working
            import numpy as np
            _ = list(model.transcribe(
                np.zeros(1600, dtype=np.float32),
                beam_size=BEAM_SIZE,
                language=LANGUAGE,
            )[0])

            using_gpu = True
            log(f"Model loaded on GPU in {load_time:.2f}s (RTF ~0.01x)")
            emit({"init": "ready", "model": MODEL_NAME, "device": "cuda",
                  "compute": COMPUTE_TYPE, "load_time_s": round(load_time, 2)})
            return

        except Exception as e:
            log(f"GPU load failed: {e}")
            log("Falling back to CPU...")

    # CPU fallback
    try:
        log(f"Loading model '{MODEL_NAME}' on CPU (int8)...")
        t0 = time.time()
        model = WhisperModel(MODEL_NAME, device="cpu", compute_type="int8")
        load_time = time.time() - t0
        device = "cpu"
        compute_type = "int8"
        log(f"Model loaded on CPU in {load_time:.2f}s")
        emit({"init": "ready", "model": MODEL_NAME, "device": "cpu",
              "compute": "int8", "load_time_s": round(load_time, 2)})
    except Exception as e:
        log(f"CPU load also failed: {e}")
        emit({"init": "error", "error": str(e)})
        raise

def get_gpu_vram():
    """Get current GPU VRAM usage in MB (if available)."""
    try:
        import subprocess
        r = subprocess.run(
            ["nvidia-smi", "--query-gpu=memory.used",
             "--format=csv,noheader,nounits"],
            capture_output=True, text=True, timeout=3
        )
        return int(r.stdout.strip())
    except:
        return None

def handle_transcribe(path: str, stream: bool = False):
    """Transcribe an audio file. If stream=True, emit partial results."""
    if not path or not os.path.exists(path):
        emit({"ok": False, "error": f"file not found: {path}"})
        return

    try:
        if stream:
            # Streaming mode: use chunk_length for incremental results
            kwargs = dict(
                beam_size=BEAM_SIZE,
                language=LANGUAGE,
                vad_filter=True,
                chunk_length=STREAM_CHUNK_LENGTH,
            )
        else:
            kwargs = dict(
                beam_size=BEAM_SIZE,
                language=LANGUAGE,
                vad_filter=True,
            )

        segments_iter, info = model.transcribe(path, **kwargs)

        if stream:
            # Emit partial results as segments arrive
            for seg in segments_iter:
                emit({"ok": True, "text": seg.text.strip(), "partial": True,
                      "start": round(seg.start, 2), "end": round(seg.end, 2)})
            # Final — join all segments
            # (Already emitted above, just signal completion)
            emit({"ok": True, "text": "", "partial": False, "done": True})
        else:
            # Batch mode: collect all segments, emit once
            text = " ".join(seg.text for seg in segments_iter).strip()
            emit({"ok": True, "text": text, "partial": False})

    except Exception as e:
        log(f"Transcribe error: {e}")
        emit({"ok": False, "error": str(e)})

def handle_health():
    vram = get_gpu_vram()
    emit({
        "ok": True,
        "healthy": True,
        "gpu": using_gpu,
        "device": device if 'device' in dir() else DEVICE,
        "model": MODEL_NAME,
        "vram_mb": vram,
    })

def main():
    try:
        load_model()
    except Exception as e:
        traceback.print_exc(file=sys.stderr)
        sys.exit(1)

    log(f"Listening for commands (model={MODEL_NAME}, device={'cuda' if using_gpu else 'cpu'})...")

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        try:
            msg = json.loads(line)
        except json.JSONDecodeError:
            emit({"ok": False, "error": "invalid json"})
            continue

        op = msg.get("op", "")

        if op == "ping":
            emit({"ok": True, "pong": True, "model": MODEL_NAME})

        elif op == "transcribe":
            handle_transcribe(msg.get("path", ""), stream=False)

        elif op == "transcribe_stream":
            handle_transcribe(msg.get("path", ""), stream=True)

        elif op == "health":
            handle_health()

        elif op == "quit":
            emit({"ok": True, "quit": True})
            break

        elif op == "set_model":
            # Hot-swap model (future use)
            new_model = msg.get("model", MODEL_NAME)
            if new_model != MODEL_NAME:
                log(f"Hot-swapping model to '{new_model}'...")
                # Just relaunch — too complex to hot-swap in-place
                MODEL_NAME = new_model
                load_model()
            else:
                emit({"ok": True, "model": MODEL_NAME})

        else:
            emit({"ok": False, "error": f"unknown op: {op}"})

if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        pass
    except Exception as e:
        traceback.print_exc(file=sys.stderr)
        print(json.dumps({"ok": False, "error": str(e)}), flush=True)
        sys.exit(1)

File: src/main/whisper-daemon-v2.ts
/**
 * Whisper Daemon V2 — Manages a persistent Python process with
 * preloaded faster-whisper model for instant transcription.
 *
 * Uses whisper_daemon_v2.py which adds:
 * - GPU support (CUDA via ctranslate2)
 * - Streaming partial transcripts
 * - Health check
 * - CPU fallback if GPU fails
 *
 * Eliminates 1-2s model-load latency per STT call by loading
 * the model once at app startup and keeping it alive.
 */

import { spawn, ChildProcess } from 'child_process';
import path from 'path';
import fs from 'fs';
import os from 'os';
import { app } from 'electron';

let daemon: ChildProcess | null = null;
let daemonReady = false;
let usingGPU = false;
let pendingQueue: Array<{
  resolve: (text: string) => void;
  reject: (err: Error) => void;
  filePath: string;
  stream: boolean;
}> = [];
let outputBuffer = '';

const hermesAgentVenvPython = path.join(
  os.homedir(), 'AppData', 'Local', 'hermes', 'hermes-agent', 'venv', 'Scripts', 'python.exe'
);

function resolveAudioScript(scriptName: string): string {
  const candidateBases = [
    path.join(__dirname, '..', 'src', 'audio'),
    path.join(__dirname, '..', '..', 'src', 'audio'),
    path.join(__dirname, '..', '..', '..', 'src', 'audio'),
    path.join(app.getAppPath(), 'src', 'audio'),
    path.join(app.getAppPath(), '..', 'src', 'audio'),
    path.join(app.getAppPath(), '..', '..', 'src', 'audio'),
  ];
  for (const base of candidateBases) {
    const candidate = path.join(base, scriptName);
    if (fs.existsSync(candidate)) return candidate;
  }
  return '';
}

export function startWhisperDaemon(): void {
  if (daemon && !daemon.killed) return;

  const scriptPath = resolveAudioScript('whisper_daemon_v2.py') || resolveAudioScript('whisper_daemon.py');
  if (!scriptPath) {
    console.error('[WhisperDaemonV2] Could not find daemon script');
    return;
  }

  const isV2 = scriptPath.includes('whisper_daemon_v2.py');
  console.log(`[WhisperDaemonV2] Starting ${isV2 ? 'V2 (GPU)' : 'V1 (CPU)'}: ${scriptPath}`);

  // Set CUDA library paths in the environment
  const venvNvidia = path.join(
    os.homedir(), 'AppData', 'Local', 'hermes', 'hermes-agent',
    'venv', 'Lib', 'site-packages', 'nvidia'
  );
  const env = { ...process.env };
  for (const sub of ['cublas', 'cudnn', 'cuda_runtime']) {
    const bindir = path.join(venvNvidia, sub, 'bin');
    if (fs.existsSync(bindir)) {
      env.PATH = bindir + (env.PATH ? path.delimiter + env.PATH : '');
    }
  }
  // Request GPU by default
  if (!env.WHISPER_DAEMON_DEVICE) env.WHISPER_DAEMON_DEVICE = 'cuda';
  if (!env.WHISPER_DAEMON_MODEL) env.WHISPER_DAEMON_MODEL = 'base';

  daemon = spawn(hermesAgentVenvPython, [scriptPath], {
    stdio: ['pipe', 'pipe', 'pipe'],
    env,
  });

  daemon.stdout?.on('data', (data: Buffer) => {
    outputBuffer += data.toString();
    const lines = outputBuffer.split('\n');
    outputBuffer = lines.pop() || ''; // keep incomplete last line

    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const msg = JSON.parse(line);
        handleDaemonMessage(msg);
      } catch {
        // non-JSON output (debug, etc.) — ignore
      }
    }
  });

  daemon.stderr?.on('data', (data: Buffer) => {
    console.error('[WhisperDaemonV2] stderr:', data.toString().trim());
  });

  daemon.on('error', (err) => {
    console.error('[WhisperDaemonV2] Process error:', err.message);
    daemonReady = false;
    rejectAll(new Error(`Whisper daemon error: ${err.message}`));
    daemon = null;
  });

  daemon.on('close', (code) => {
    console.log('[WhisperDaemonV2] Exited with code:', code);
    daemonReady = false;
    rejectAll(new Error('Whisper daemon exited'));
    daemon = null;
  });
}

function handleDaemonMessage(msg: any) {
  if (msg.init === 'ready') {
    daemonReady = true;
    usingGPU = msg.device === 'cuda';
    console.log(
      `[WhisperDaemonV2] Ready — model=${msg.model}, device=${msg.device}, ` +
      `compute=${msg.compute}, load=${msg.load_time_s}s`
    );
    flushPending();
  } else if (msg.init === 'error') {
    console.error('[WhisperDaemonV2] Init error:', msg.error);
    rejectAll(new Error(msg.error));
  } else if (msg.ok === true && msg.text !== undefined) {
    // Transcription result (batch or streaming partial)
    const next = pendingQueue[0];
    if (next) {
      if (msg.partial) {
        // Streaming partial — don't resolve yet, just log
        console.log(`[WhisperDaemonV2] Partial: "${msg.text.substring(0, 40)}..."`);
      } else {
        // Final result
        pendingQueue.shift();
        next.resolve(msg.text);
      }
    }
  } else if (msg.ok === true && msg.pong === true) {
    // Ping response
    console.log(`[WhisperDaemonV2] Ping OK — model=${msg.model}`);
  } else if (msg.ok === true && msg.healthy === true) {
    // Health check response
    console.log(
      `[WhisperDaemonV2] Health: gpu=${msg.gpu}, device=${msg.device}, ` +
      `vram=${msg.vram_mb}MB`
    );
  } else if (msg.ok === false && msg.error) {
    const next = pendingQueue.shift();
    if (next) next.reject(new Error(msg.error));
  }
}

export function stopWhisperDaemon(): void {
  if (!daemon || daemon.killed) return;
  try {
    daemon.stdin?.write(JSON.stringify({ op: 'quit' }) + '\n');
  } catch {}
  setTimeout(() => {
    if (daemon && !daemon.killed) daemon.kill();
  }, 2000);
}

/**
 * Transcribe a file via the daemon (batch mode).
 */
export function transcribeViaDaemon(filePath: string): Promise<string> {
  return enqueueTranscribe(filePath, false);
}

/**
 * Transcribe a file via the daemon with streaming partial results.
 * Partial results are emitted to the callback as they arrive.
 */
export function transcribeStreamViaDaemon(
  filePath: string,
  onPartial: (text: string) => void
): Promise<string> {
  // For now, same as batch — streaming partials arrive as stdout messages
  // but we resolve on the final. Partial callback is a future enhancement.
  return enqueueTranscribe(filePath, true);
}

function enqueueTranscribe(filePath: string, stream: boolean): Promise<string> {
  return new Promise((resolve, reject) => {
    pendingQueue.push({ resolve, reject, filePath, stream });

    if (!daemon || daemon.killed) {
      startWhisperDaemon();
      return; // Will be processed when ready
    }

    if (daemonReady) {
      const op = stream ? 'transcribe_stream' : 'transcribe';
      try {
        daemon.stdin?.write(JSON.stringify({ op, path: filePath }) + '\n');
      } catch (e) {
        const idx = pendingQueue.findIndex(q => q.resolve === resolve);
        if (idx >= 0) pendingQueue.splice(idx, 1);
        reject(new Error('Failed to send to daemon'));
      }
    }
  });
}

export function isWhisperDaemonReady(): boolean {
  return daemonReady && daemon !== null && !daemon!.killed;
}

export function isWhisperUsingGPU(): boolean {
  return usingGPU;
}

export function pingDaemon(): void {
  if (daemon && daemonReady) {
    try {
      daemon.stdin?.write(JSON.stringify({ op: 'ping' }) + '\n');
    } catch {}
  }
}

/**
 * Request a health check from the daemon.
 */
export function healthCheckDaemon(): void {
  if (daemon && daemonReady) {
    try {
      daemon.stdin?.write(JSON.stringify({ op: 'health' }) + '\n');
    } catch {}
  }
}

function flushPending(): void {
  for (const q of pendingQueue) {
    try {
      const op = q.stream ? 'transcribe_stream' : 'transcribe';
      daemon?.stdin?.write(JSON.stringify({ op, path: q.filePath }) + '\n');
    } catch {
      // Will handle in error/close
    }
  }
}

function rejectAll(err: Error): void {
  const remaining = [...pendingQueue];
  pendingQueue = [];
  for (const q of remaining) {
    q.reject(err);
  }
}
